Update `deserialize_witness` fuzz target to use the `Witness` `arbitrary` impl

shinghim opened the pull request:
The result should be that the roundtrip logic is tested more often due to `arbitrary` yielding more valid Witnesses versus the previous case, which attempted to deserialize random bytes into a witness. IIUC this should also allow the fuzzer to find edge cases more easily as well

apoelstra commented:
Concept ACK. Though this particular target should be renamed `parse_amount` from `deserialize_amount`, and left alone since its point is to fuzz the string parsing logic rather than the amount construction logic.

It should also be changed to read a control byte rather than attempting to parse a signed amount followed by an unsigned..

Basically, great idea but pick a different fuzz target :).

apoelstra commented on fuzz/fuzz_targets/bitcoin/deserialize_witness.rs:
```diff
@@ -1,12 +1,19 @@
-        assert_eq!(data, &serialized[..]);
+        let deserialized: Result<Witness, _> = deserialize(serialized.as_slice());
+
+        if let Ok(deserialized_witness) = deserialized {
```
In 17e899c425f6493eec1f3e2e96885bcf6cb48ff2:

We should assert the result of `deserialize`. We should be able to deserialize anything that we serialized that we get out of `arbitrary`. (I think, in general we can construct massive witnesses that we'll serialize and then refuse to deserialize. Though I'm unsure. In any case I don't think our arbitrary impl should be producing such things.)

I added an unwrap and fuzzed for 15 million iterations on my machine as a quick sanity check of this.

shinghim replied:
Sounds good. There is a way to cap the size of the vectors if it becomes an issue

apoelstra reviewed (dismissed):
ACK 3e94d8206fb0f17c29392af910db26f7207a1b70; successfully ran local tests

shinghim commented:
Updated the first commit with the snippet you posted. I generated the files again, and it moved one more of the fuzz targets, but the change that moved `bitcoin_deser_net_msg` to the top of the list stuck.

I think what might have happened is that there were some edits made manually to the files that `generate-files.sh` updates, because there were other changes to `fuzz/Cargo.toml` that I didn't add to the template in `fuzz/generate-files.sh` under `# 1. Generate fuzz/Cargo.toml` such as the `rust-version` field. I also added the `p2p` dependency to the `fuzz/generate-files.sh` template, but it was already in `fuzz/Cargo.toml`

apoelstra commented:
>Is this because you'd like the targets to be fuzzed in a particular order, or does this have deeper implications on how the fuzzer is run?

No, it's just that I check these things with my local CI and it should fail.

The fact that it didn't is a little concerning actually.. will dig into that.
